pkg/operator/ceph/object/user/controller.go

```diff
@@ -151,7 +155,63 @@ func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cep
 		DisplayName: &displayName,
 	}
 	objContext := object.NewContext(context, u.Spec.Store, u.Namespace)
-
+	// Wait with exponential backoff for the objectstore to be created
+	//
+	// ExponentialBackoff repeats a condition check with exponential backoff
+	//
+	// If the condition never returns true, ErrWaitTimeout is returned
+	// All errors except IsNotFound terminates immediately
+	//
+	// On Step 1 we'll wait for 15 sec (i.e. Duration)
+	// On Step 2 we'll wait for 15*1.5=22.5 sec (i.e. Previous Duration * Factor)
+	// On Step 3 we'll wait for 15*(1.5^2)=33.75 sec (i.e. Duration * (Factor^(Current Step-1)) )
+	// Once we have this duration > Cap (i.e 5 mins) or Step=10, we stop waiting
+	//
+	// In this configuration, we will wait for 5 mins making 8 retries with exponential backoff
+	err := wait.ExponentialBackoff(wait.Backoff{
```

Comment:
I think this exponential backoff is potentially dangerous for Rook. We are wanting to wait some number of minutes for a case to be true, but with the exponential backoff, this seems to be implemented under the hood with a dumb timeout, and the condition is not checked one last time after the timeout. In the example I made to test this more quickly, I want to wait 10 seconds for a condition, but it only actually tests it up to 5.0625 seconds. This seems to me to defeat what we are trying to achieve in Rook.

And based on my math, this configuration will only test up to 256.2891 seconds, which is over 30 seconds shorter than the desired 5 minute timeout. And because of the exponential equation, this becomes difficult to reason about. 

Why don't we either (A) use a linear timeout that we can reason about more easily or (B) use a watcher that will put less pressure on the API server?

Reply:
It was my suggestion to use the exponential backoff so you can blame me :). In many places K8s is using a backoff scheme to ensure the api server is not being overwhelmed. That said, it does seem to be more complexity than we need and causes all these questions around the behavior. 

If we're using a relatively large polling interval, it would certainly be simpler to follow what we're doing other places and just query with the same interval with `wait.Poll()` and keep it simple. As long as it's only polling from a single place (the operator) rather than all pods, it's not going to overwhelm the API server.  What if we just poll every 15 seconds for 5 minutes? Nice and simple. 

@umangachapagain What do you think? Thanks for your patience while we talk through whether a new pattern is the right thing...

Reply:
> implemented under the hood with a dumb timeout 

Can you explain this? The way I see it, we have a set of parameters that we can configure (including the dumb timeout). Based on these parameters, a predefined mathematical function is used to create a "Wait".

> and the condition is not checked one last time after the timeout

I believe this is literally what a "timeout" is supposed to do.

> this configuration will only test up to 256.2891 seconds, which is over 30 seconds shorter than the desired 5 minute timeout

we could easily add "30 sec" to Cap and get back the lost seconds. If that seems problematic.

> use a linear timeout that we can reason about more easily

Linear or Exponential both are reasonable mathematical functions. And if my understanding is correct Exponential Timeout was designed to better handle the server load without always making the same assumptions (Polling every 10 or 15 second even after knowing that Pod could take more than a minute to come up, wasting at least 4 retries).

> What if we just poll every 15 seconds for 5 minutes?

If this is desired, setting the Factor to 1 will give you the same effect.

Here is my reasoning for why Exponential Backoff is good, based on the configuration I have used,
1. First wait (15 sec) is ideal to wait for the ObjectStore CR to be created after ObjectUser Cr is created. If for any reason this does not happen, Second Wait (22.5 sec) for Objectstore CR to appear making it a total wait of 37.5 sec for both CRs to be applied sounds reasonable to me.
2. After this happens, third & fourth wait (totaling >1 min) should be enough for a rgw pod to be running, unless there is a major performance issue.
3. So, by 4th wait we should be good and exiting the wait loop. If by any chance that doesn't happen, 5th 6th 7th 8th waits (each wait between 1.2 mins - 4.2mins) should be more than sufficient for all the required resources to come up and we should be good to proceed without having to worry about the 30sec lost.

That being said, if you still think linear wait or watch is better, I will happily abandon this work. This is a good discussion and so much to learn :)

Reply:
The exponential backoff will certainly work, there is no doubt. Here are the downsides to it that I'm seeing:
- Developer complexity/code readability is more difficult. It's just harder to understand.
- There isn't really a need for it. The need for exponential backoff is when there are many clients that would be flooding the server. In this case, it's just a single client and the api server will not be overloaded by this request.
- As the backoff wait time increases with each try, you have the potential for waiting longer unnecessarily to create the resource after it's ready to be created. A linear approach of 15s will likely create the resource sooner.